KEYCLOAK-6160 Docs for X509 certificate lookup provider

mposolda opened the pull request:
- Also fixed some occurences of hardcoded stuff like "Keycloak" to use expressions.

mposolda committed:
KEYCLOAK-6160 Docs for X509 certificate lookup provider

matthewhelmke commented on server_admin/topics/authentication/x509.adoc:
```diff
@@ -209,6 +209,61 @@ image:images/x509-directgrant-flow.png[]
 
+==== Client certificate lookup
+
+When user sends HTTP requests directly to {project_name} server, it's the {appserver_name} undertow subsystem, which will establish the SSL handshake and extracts client certificate. Client certificate will be then saved to the attribute `javax.servlet.request.X509Certificate` of the HTTP request as specified in the servlet specification. {project_name} X509 authenticator will be then able to loo ...
```
When an HTTP request is sent directly to {project_name} server, the {appserver_name} undertow subsystem will establish an SSL handshake and extract the client certificate. The client certificate will be then saved to the attribute `javax.servlet.request.X509Certificate` of the HTTP request, as specified in the servlet specification. The {project_name} X509 authenticator will be then able to lookup the certificate from this attribute.

matthewhelmke commented on server_admin/topics/authentication/x509.adoc:
```diff
@@ -209,6 +209,61 @@ image:images/x509-directgrant-flow.png[]
+
+When user sends HTTP requests directly to {project_name} server, it's the {appserver_name} undertow subsystem, which will establish the SSL handshake and extracts client certificate. Client certificate will be then saved to the attribute `javax.servlet.request.X509Certificate` of the HTTP request as specified in the servlet specification. {project_name} X509 authenticator will be then able to loo ...
+
+However when {project_name} listens to HTTP requests behind the loadbalancer or reverse proxy, it might be the proxy server, which extracts the client certificate and establish the mutual SSL connection. Reverse proxy usually put the authenticated client certificate to the HTTP header of the underlying request and forwards it to the backend {project_name} server. In this case, {project_name} must ...
```
However, when the {project_name} server listens to HTTP requests behind a load balancer or reverse proxy, it may be the proxy server which extracts the client certificate and establishes the mutual SSL connection. A reverse proxy usually puts the authenticated client certificate in the HTTP header of the underlying request and forwards it to the back end {project_name} server. In this case, {project_name} must be able to look up the X.509 certificate chain from the HTTP headers instead of from the attribute of HTTP request, as is done for Undertow.

matthewhelmke commented on server_admin/topics/authentication/x509.adoc:
```diff
@@ -209,6 +209,61 @@ image:images/x509-directgrant-flow.png[]
+
+However when {project_name} listens to HTTP requests behind the loadbalancer or reverse proxy, it might be the proxy server, which extracts the client certificate and establish the mutual SSL connection. Reverse proxy usually put the authenticated client certificate to the HTTP header of the underlying request and forwards it to the backend {project_name} server. In this case, {project_name} must ...
+
+If {project_name} is behind the reverse proxy, you usually need to configure alternative provider of the `x509cert-lookup` SPI in `{project_dirref}/standalone/configuration/standalone.xml` file. Besides the `default` provider, which lookups the certificate from the HTTP header, we have 2 builtin providers: `haproxy` and `apache`.
```
If {project_name} is behind a reverse proxy, you usually need to configure alternative provider of the `x509cert-lookup` SPI in `{project_dirref}/standalone/configuration/standalone.xml`. Along with the `default` provider, which looks up the certificate from the HTTP header, we also have two additional built-in providers: `haproxy` and `apache`, which are described next.

matthewhelmke commented on server_admin/topics/authentication/x509.adoc:
```diff
@@ -209,6 +209,61 @@ image:images/x509-directgrant-flow.png[]
+
+===== HAProxy certificate lookup provider
+
+This provider can be used when your {project_name} server is behind the HAProxy reverse proxy. The provider needs to be configured like this:
```
You can use this provider when your {project_name} server is behind an HAProxy reverse proxy. Configure the server like this:

matthewhelmke commented on server_admin/topics/authentication/x509.adoc:
```diff
@@ -209,6 +209,61 @@ image:images/x509-directgrant-flow.png[]
+
+In this example configuration, the client certificate will be lookup from the HTTP header `SSL_CLIENT_CERT` and the other certificates from it's chain will be lookup from the HTTP headers like `CERT_CHAIN_0` , `CERT_CHAIN_1`, ..., `CERT_CHAIN_9` . The attribute `certificateChainLength` is the maximum length of the chain, so the last one tried attribute would be `CERT_CHAIN_9` .
+
+Consult the HAProxy documentation for the details how the HTTP Headers for the client certificate and client certificate chain can be configured and what are their proper names.
```
In this example configuration, the client certificate will be looked up from the HTTP header, `SSL_CLIENT_CERT`, and the other certificates from its chain will be looked up from HTTP headers like `CERT_CHAIN_0` , `CERT_CHAIN_1`, ..., `CERT_CHAIN_9` . The attribute `certificateChainLength` is the maximum length of the chain, so the last one tried attribute would be `CERT_CHAIN_9` .

Consult the HAProxy documentation for the details of how the HTTP Headers for the client certificate and client certificate chain can be configured and their proper names.

matthewhelmke commented on server_admin/topics/authentication/x509.adoc:
```diff
@@ -209,6 +209,61 @@ image:images/x509-directgrant-flow.png[]
+
+===== Apache certificate lookup provider
+
+This provider can be used when your {project_name} server is behind the Apache reverse proxy. The provider needs to be configured like this:
```
You can use this provider when your {project_name} server is behind an Apache reverse proxy. Configure the server like this:

matthewhelmke commented on server_admin/topics/authentication/x509.adoc:
```diff
@@ -209,6 +209,61 @@ image:images/x509-directgrant-flow.png[]
+</spi>
+----
+
+The configuration is same like for `haproxy` provider. Again, consult the Apache documentation for the details how the HTTP Headers for the client certificate and client certificate chain can be configured and what are their names.
```
The configuration is same as for the `haproxy` provider. Consult the Apache documentation for the details of how the HTTP Headers for the client certificate and client certificate chain can be configured and their proper names.

matthewhelmke commented on server_admin/topics/authentication/x509.adoc:
```diff
@@ -209,6 +209,61 @@ image:images/x509-directgrant-flow.png[]
+
+We don't have builtin support for the other reverse proxy implementations. However it's possible that other reverse proxies can behave
+in similar way like `apache` or `haproxy` and some of those providers can be used. Alternatively you may need to create your own implementation of the `org.keycloak.services.x509.X509ClientCertificateLookupFactory` and `org.keycloak.services.x509.X509ClientCertificateLookup` provider. See the link:{developerguide_link}[{developerguide_name}] for the details how to add your own provider.
+
```
We do not have built-in support for other reverse proxy implementations. However, it is possible that other reverse proxies can be made to behave in a similar way to `apache` or `haproxy` and that some of those providers can be used. If you want to try, you may need to create your own implementation of the `org.keycloak.services.x509.X509ClientCertificateLookupFactory` and `org.keycloak.services.x509.X509ClientCertificateLookup` provider. See the link:{developerguide_link}[{developerguide_name}] for the details on how to add your own provider.

matthewhelmke reviewed (changes requested):
Looks pretty good. I cleaned up the language and left my rewritten suggestions in the comments. If you are content, you can just cut/paste them in.

mposolda commented:
Thanks @matthewhelmke ! I've changed it per your feedback. 

Only small exception is the sentence in last paragraph, which I changed to: "If none of those works, you may need to ..." instead of "If you want to try, you may need ..." . This was to emphasize, that you need to create your own provider just in the case that none of the built-in providers can be used.

matthewhelmke commented:
Thanks, @mposolda. Looks good! I like what you chose to do on that last sentence, too.

mposolda commented:
@matthewhelmke Thanks!
